### DISCUSSION TITLE
when-asp-net-core-cant-find-your-controller-debugging-application-parts/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#130] [Announcements]
# when-asp-net-core-cant-find-your-controller-debugging-application-parts/

In this post I describe application parts, how ASP.NET Core uses them to find the controllers in your app, and how to list them.

https://andrewlock.net/when-asp-net-core-cant-find-your-controller-debugging-application-parts/

**Comment by zwoolli** [comment]
Thanks for the article! I have moved my api controllers, razor pages, and view components to a `Presentation` project, and I have moved all Identity Framework logic (Authentication / Authorization) to a `Service` project. I am struggling to get the application to run. I followed what you have outlined in this article and my controller is indeed showing up as an application part. How can I do the same with razor pages and view components? I am not seeing a RazorPageFeature in the source code.

**Reply by zwoolli** [reply]
I ended up using the [`ViewFeature`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razor.compilation.viewsfeature?view=aspnetcore-7.0) class and following your lead with the `ControllerFeature` code. As it turned out, none of my Razor Pages were seen by my Web app. 
```
...
var viewFeature = new ViewsFeature();
 _partManager.PopulateFeature(viewFeature);
var views = viewFeature.ViewDescriptors.Select(x => x.Item?.Identifier);

// Log the application parts and controllers
_logger.LogInformation("Found the following application parts: '{ApplicationParts}' with the following controllers: '{Controllers}' and views '{Views}'",
        string.Join(", ", applicationParts), string.Join(", ", controllers), string.Join(", ", views));
...
```
I changed my `Presentation` project from a regular class library project to a [Razor class library project](https://learn.microsoft.com/en-us/aspnet/core/razor-pages/ui-class?view=aspnetcore-7.0&tabs=visual-studio) and that did the trick. I still do not know how to make the web app see a regular class library containing Razor pages and controllers 🤷🏽‍♀️. But for now, everything is working. Thanks again for your article 😊

[a comment by an automated account is not shown]

**Comment by kwaazaar** [comment]
I don't see the assembly-level `ApplicationPartAttribute` when I inspect the dll with dotPeek. Is this still happening this way in .NET 8? (the examples and documentation all mention .NET Core 3)
I ask because *sometimes* while debugging my application only the controllers in the main project work, not the controllers from class library projects. This does not happen all the time though. I tried to create a separate repro, but of course that one always works :-s

**Comment by mattermani** [comment]
Great article!!! Still relevant and I used ' .AddApplicationPart(typeof(TestController).Assembly)' to host a webapi from my integration test. Simply amazing, dont think I could have had that working without this article. Thank you Andrew ! I am definitely getting your latest book :)
